pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java

```diff
@@ -0,0 +1,364 @@
+    return sendPutRequest(uri, payload, headers, null);
+  }
+
+  public SimpleHttpResponse sendPutRequest(URI uri, HttpEntity payload, Map<String, String> headers,
+      @Nullable String authToken)
+      throws IOException {
+    RequestBuilder requestBuilder = RequestBuilder.put(uri).setVersion(HttpVersion.HTTP_1_1);
+    if (payload != null) {
+      requestBuilder.setEntity(payload);
+    }
+    if (StringUtils.isNotBlank(authToken)) {
+      requestBuilder.addHeader(AUTH_HTTP_HEADER, authToken);
+    }
+    setTimeout(requestBuilder, DELETE_REQUEST_SOCKET_TIMEOUT_MS);
+    return sendRequest(requestBuilder.build());
+  }
+
+  // --------------------------------------------------------------------------
+  // JSON post/put utility APIs
+  // --------------------------------------------------------------------------
+
+  public SimpleHttpResponse postJsonRequest(URI uri, @Nullable String jsonRequestBody)
+      throws IOException {
+    return postJsonRequest(uri, jsonRequestBody, null);
+  }
+
+  public SimpleHttpResponse postJsonRequest(URI uri, @Nullable String requestBody,
+      @Nullable Map<String, String> headers)
+      throws IOException {
+    if (MapUtils.isEmpty(headers)) {
+      headers = new HashMap<>();
+    }
+    headers.put(HttpHeaders.CONTENT_TYPE, JSON_CONTENT_TYPE);
+    HttpEntity entity = requestBody == null ? null : new StringEntity(requestBody, ContentType.APPLICATION_JSON);
+    return sendPostRequest(uri, entity, headers, null);
+  }
+
+  public SimpleHttpResponse putJsonRequest(URI uri, @Nullable String jsonRequestBody)
+      throws IOException {
+    return putJsonRequest(uri, jsonRequestBody, null);
+  }
+
+  public SimpleHttpResponse putJsonRequest(URI uri, @Nullable String requestBody,
+      @Nullable Map<String, String> headers)
+      throws IOException {
+    if (MapUtils.isEmpty(headers)) {
+      headers = new HashMap<>();
+    }
+    headers.put(HttpHeaders.CONTENT_TYPE, JSON_CONTENT_TYPE);
+    HttpEntity entity = requestBody == null ? null : new StringEntity(requestBody, ContentType.APPLICATION_JSON);
+    return sendPutRequest(uri, entity, headers, null);
+  }
+
+  // --------------------------------------------------------------------------
+  // Lower-level request/execute APIs.
+  // --------------------------------------------------------------------------
+
+  public SimpleHttpResponse sendRequest(HttpUriRequest request)
+      throws IOException {
+    try (CloseableHttpResponse response = _httpClient.execute(request)) {
+      if (response.containsHeader(CommonConstants.Controller.HOST_HTTP_HEADER)) {
+        String controllerHost = response.getFirstHeader(CommonConstants.Controller.HOST_HTTP_HEADER).getValue();
+        String controllerVersion = response.getFirstHeader(CommonConstants.Controller.VERSION_HTTP_HEADER).getValue();
+        LOGGER.info(String
+            .format("Sending request: %s to controller: %s, version: %s", request.getURI(), controllerHost,
```

Comment:
Just for my own edification, I thought that since we were using slf4j, the approved approach for formatted log msgs was to use slf4j's support for `{}` parameters. E.g.
``` java
LOGGER.info("Sending request: {} to controller: {}, version: {}", request.getURI(), controllerHost, controllerVersion);
```

Reply:
Yes, please do not use `String.format` - it's ridiculously expensive. Most of the time we use it, it is when something has gone wrong so I suppose it is ok but it should never be used on the happy path.

Reply:
In this particular case, since there are 3 parameters a varargs array will be allocated for every log statement which will always be logged since it's at INFO, so it would be better just to concatenate the strings to minimise logging overhead.